system/lib/wasmfs/wasmfs.cpp

```diff
@@ -8,60 +8,258 @@
-          if (fd == 1) {
-            emscripten_console_log(&fd_write_stdstream_buffer[0]);
-          } else if (fd == 2) {
-            emscripten_console_error(&fd_write_stdstream_buffer[0]);
-          }
-          fd_write_stdstream_buffer.clear();
-        } else {
-          fd_write_stdstream_buffer.push_back(current);
-        }
+  __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) override {
+    return write_buffer(iovs, iovs_len, nwritten, &emscripten_console_error);
+  }
+};
+
+class OpenFile {
+  std::shared_ptr<File> file_node;
+  uint32_t _offset;
+  std::mutex openfile_mutex;
+
+public:
+  OpenFile(uint32_t offset, std::shared_ptr<File> _file_node) {
+    this->_offset = offset;
+    this->file_node = _file_node;
+  }
+
+  std::shared_ptr<File> get_file_node() { return file_node; }
+};
+
+class FileTable {
+  std::vector<std::shared_ptr<OpenFile>> handles;
+  std::mutex ft_mutex;
+
+public:
+  std::vector<std::shared_ptr<OpenFile>> get_file_handles() { return handles; }
+  std::shared_ptr<OpenFile> get_open_file(__wasi_fd_t fd) {
+    // Check if the file descriptor is invalid.
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd >= handles.size() || fd < 0)
+      return nullptr;
+
+    return handles[fd];
+  }
+
+  __wasi_fd_t add_open_file(std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    for (__wasi_fd_t i = 0; i < handles.size(); i++) {
+      if (!handles[i]) {
+        // Free open file entry.
+        handles[i] = ptr;
+        return i;
       }
-      num += len;
     }
-    *nwritten = num;
+
+    // Could not find an empty open file table entry.
+    handles.push_back(ptr);
+
+    return handles.size() - 1;
   }
-  return 0;
+
+  int remove_open_file(__wasi_fd_t fd) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    // Check if the file descriptor is invalid.
+    if (fd >= handles.size() || fd < 0)
+      return -1;
+
+    handles[fd] = nullptr;
+    return 0;
+  }
+
+  int set_open_file(__wasi_fd_t fd, std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd < 0)
+      return -1;
+
+    if (fd >= handles.size()) {
+      handles.resize(fd);
+      handles.insert(handles.begin() + fd, ptr);
+    } else {
+      handles[fd] = ptr;
+    }
+
+    return 0;
+  }
+};
+
+// Global filetable singleton
+static FileTable* ft = nullptr;
+
+void populate_file_table() {
+  ft = new FileTable();
+  ft->add_open_file(std::make_shared<OpenFile>(0, std::make_shared<StdinFile>()));
+  ft->add_open_file(std::make_shared<OpenFile>(0, std::make_shared<StdoutFile>()));
+  ft->add_open_file(std::make_shared<OpenFile>(0, std::make_shared<StderrFile>()));
+}
+
+__wasi_fd_t __syscall63(__wasi_fd_t oldfd, __wasi_fd_t newfd) { // dup2
+  if (newfd < 0) {
+    return -1;
+  }
+  if (!ft) {
+    populate_file_table();
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
```

Comment:
This check needs to happen after the check that `oldfd` is a valid file descriptor. It would be good to have tests for these corner cases in the `dup2` semantics.

Reply:
Yep, will add a test for this case!